### PR TITLE
Multi-az support

### PR DESCRIPTION
**Pull request by cedbossneo**

### CODE REVIEW

File: multi-node/aws/build [hunk 1]
```diff
@@ -15,7 +15,8 @@ if [ -n "$(git diff --shortstat 2> /dev/null | tail -n1)" ]; then
 fi
 
 
-
+echo "Restoring deps"
+godep restore
```
Comment (bcwaldon) [inline multi-node/aws/build]: This should not be necessary. The `go build` command below overrides `GOPATH` to place the workspace earlier in the search path than your existing `GOPATH`.

File: multi-node/aws/.gitignore [hunk 2]
```diff
@@ -2,3 +2,4 @@
 /bin
 /clusters
 /cluster.yaml
+Godeps/_workspace
```
Comment (bcwaldon) [inline multi-node/aws/.gitignore]: This is not necessary

File: multi-node/aws/pkg/cluster/cloud_config_util.go [hunk 3]
```diff
@@ -8,7 +8,7 @@ import (
 var templateRef *regexp.Regexp
 
 func init() {
-	templateRef = regexp.MustCompile(`{{\s*[a-zA-Z0-9]+(\|[a-zA-Z0-9]+)*\s*}}`)
+	templateRef = regexp.MustCompile(`{{\s*[a-zA-Z0-9\:]+(\|[a-zA-Z0-9]+)*\s*}}`)
```
Comment (bcwaldon) [inline multi-node/aws/pkg/cluster/cloud_config_util.go]: What is this change for?

File: multi-node/aws/pkg/cluster/cloud_config_worker.go [hunk 4]
```diff
@@ -7,7 +7,7 @@ coreos:
 
   flannel:
     interface: $private_ipv4
-    etcd_endpoints: http://10.0.0.50:2379
+    etcd_endpoints: http://master.kube.internal:2379
```
Comment (bcwaldon) [inline multi-node/aws/pkg/cluster/cloud_config_worker.go]: I feel like this should be `kubernetes.cluster.internal`. @aaronlevy ?

File: multi-node/aws/pkg/cluster/stack_template.go [hunk 5]
```diff
@@ -13,19 +13,23 @@ const (
 	resNameVPCGatewayAttachment         = "VPCGatewayAttachment"
 	resNameRouteTable                   = "RouteTable"
 	resNameRouteToInternet              = "RouteToInternet"
-	resNameSubnet                       = "Subnet"
-	resNameSubnetRouteTableAssociation  = "SubnetRouteTableAssociation"
+	resNameSubnetA 						= "SubnetA"
```
Comment (bcwaldon) [inline multi-node/aws/pkg/cluster/stack_template.go]: spacing is off

File: multi-node/aws/pkg/cluster/stack_template.go [hunk 6]
```diff
@@ -13,19 +13,23 @@ const (
...
-	resNameAlarmControllerRecover       = "AlarmControllerRecover"
 	resNameAutoScaleWorker              = "AutoScaleWorker"
 	resNameLaunchConfigurationWorker    = "LaunchConfigurationWorker"
+	resNameAutoScaleController          = "AutoScaleController"
+	resNameLaunchConfigurationController= "LaunchConfigurationController"
 	resNameIAMRoleController            = "IAMRoleController"
 	resNameIAMInstanceProfileController = "IAMInstanceProfileController"
 	resNameIAMRoleWorker                = "IAMRoleWorker"
 	resNameIAMInstanceProfileWorker     = "IAMInstanceProfileWorker"
+	resNameRoute53Zone					= "VPCZone"
```
Comment (bcwaldon) [inline multi-node/aws/pkg/cluster/stack_template.go]: spacing is off

File: multi-node/aws/pkg/cluster/cluster.go [hunk 7]
```diff
@@ -156,25 +157,18 @@ func (c *Cluster) Create(tlsConfig *TLSConfig) error {
...
-		parameters = append(parameters, &cloudformation.Parameter{
-			ParameterKey:     aws.String(parAvailabilityZone),
-			ParameterValue:   aws.String(c.cfg.AvailabilityZone),
-			UsePreviousValue: aws.Bool(true),
-		})
-	}
-
 	tmplURL := fmt.Sprintf("%s/template.json", c.cfg.ArtifactURL)
-	return createStackAndWait(cloudformation.New(c.aws), c.stackName(), tmplURL, parameters)
+	return createStackAndWait(cloudformation.New(session.New(c.aws)), c.stackName(), tmplURL, parameters)
```
Comment (bcwaldon) [inline multi-node/aws/pkg/cluster/cluster.go]: Why is the added `session.New()` call necessary?

File: multi-node/aws/pkg/cluster/stack_template.go [hunk 8]
```diff
@@ -396,82 +431,86 @@ func StackTemplateBody(defaultArtifactURL string) (string, error) {
...
-			"Statistic":          "Minimum",
-			"Period":             "60",
-			"EvaluationPeriods":  "5",
-			"ComparisonOperator": "GreaterThanThreshold",
-			"Threshold":          "0",
-			"AlarmActions": []interface{}{
+			"HostedZoneConfig":  map[string]interface{}{
+				"Comment": "The Kubernetes Internal zone",
+			},
+			"Name": "kuber.internal",
```
Comment (bcwaldon) [inline multi-node/aws/pkg/cluster/stack_template.go]: typo

File: multi-node/aws/pkg/cluster/stack_template.go [hunk 9]
```diff
@@ -590,17 +669,18 @@ func StackTemplateBody(defaultArtifactURL string) (string, error) {
 		"Description": "Number of worker instances to create, may be modified later",
 	}
 
+	par[parControllerCount] = map[string]interface{}{
```
Comment (bcwaldon) [inline multi-node/aws/pkg/cluster/stack_template.go]: This needs to be configurable through cluster.yaml

File: multi-node/aws/pkg/cluster/cloud_config_util.go (hunk 3 above)
Comment (cedbossneo) [inline multi-node/aws/pkg/cluster/cloud_config_util.go]: This allow handling AWS::Region in template

File: multi-node/aws/pkg/cluster/cloud_config_worker.go (hunk 4 above)
Comment (cedbossneo) [inline multi-node/aws/pkg/cluster/cloud_config_worker.go]: This allows every node access the kubernetes master through load balancer

File: multi-node/aws/pkg/cluster/cluster.go (hunk 7 above)
Comment (cedbossneo) [inline multi-node/aws/pkg/cluster/cluster.go]: I had to upgrade deps to handle ELB and it's now necessary to create a session instance.

**Comment by cedbossneo** [comment]
Ready for another review.

File: multi-node/aws/pkg/cluster/cluster.go (hunk 7 above)
Comment (bcwaldon) [inline multi-node/aws/pkg/cluster/cluster.go]: OK, I see you updated the vendored godeps

**Comment by bcwaldon** [comment]
@cedbossneo Overall, this looks good. The one architectural change I want to suggest is to not manage the `kube.internal` zone at all, and to route traffic from workers directly to the ELB hostname. Would this be possible?

**Comment by cedbossneo** [comment]
This should not be a problem. I will make this change quickly after
Christmas. :)

**Comment by cedbossneo** [comment]
Ready for review

**Comment by aaronlevy** [comment]
Sorry I'm late to this PR. So it looks like a few things are going on here:

- Adds multiple master nodes (and ELB fronting masters)
- Adds option of 2 availability zones
- Bumps AWS SDK version
- Modifies template replacements

I feel like each of these would need to really be broken out into its own PR. As it stands this is pretty difficult to adequately review, because it touches on so much.

Some general notes:

- This doesn't seem to handle etcd replication. It would seem that a request routed through the ELB would be hitting a different (and isolated) etcd instance on each master node.

- Only supports 2 static AZs. This leads to a lot of "A/B" code duplication and I think it would be more valuable to allow launching nodes into arbitrary # of AZs without the static AZ-A/AZ-B code.

    An option here could be modifying the tool to allow you to re-run and launch "more" workers in different AZs that attach to an existing master. See upstream style support for this: https://github.com/kubernetes/kubernetes/pull/16735

- Bumping AWS SDK should be done in its own PR, and also tested in isolation. I believe the version we currently have supports ELBs, but either way, both changing the SDK and the functionality of the tool should be done separately.

- Changing the template replacements also should likely be broken out into another PR so the implications / upgradability / etc can be discussed and tested in isolation of functionality changes in the rest of the tool.

As this PR currently stands, I don't feel comfortable merging. This could be broken out into smaller units of work that are more easily tested / reasoned about. I might even suggest for the larger pieces (multi-master / multi-az) writing a proposal document so the implementation can be discussed prior to you spending a lot of time on code.

Related (automated conformance testing): https://github.com/coreos/coreos-kubernetes/issues/224

**Pull request closed without merging**